lib/services/role.go

```diff
@@ -850,49 +867,110 @@ func NewEnumerationResult() EnumerationResult {
+}
+
+func (set RoleSet) EnumerateDatabaseNames(database types.Database, extraNames ...string) EnumerationResult {
+	getter := func(role types.Role, cond types.RoleConditionType) []string {
+		return role.GetDatabaseNames(cond)
+	}
+	matchMaker := func(name string) RoleMatcher {
+		return &DatabaseNameMatcher{Name: name}
+	}
+	return set.enumerateEntities(database, getter, matchMaker, extraNames...)
+}
+
+type EntityGetter func(types.Role, types.RoleConditionType) []string
+type RoleMatchMaker func(string) RoleMatcher
+
+func (set RoleSet) enumerateEntities(r AccessCheckable, getter EntityGetter, matchMaker RoleMatchMaker, extraEntities ...string) EnumerationResult {
 	result := NewEnumerationResult()
 
-	// gather users for checking from the roles, check wildcards.
-	var users []string
+	// gather entities for checking from the roles, check wildcards.
+	var entities []string
 	for _, role := range set {
 		wildcardAllowed := false
 		wildcardDenied := false
 
-		for _, user := range role.GetDatabaseUsers(types.Allow) {
-			if user == types.Wildcard {
+		for _, entity := range getter(role, types.Allow) {
+			if entity == types.Wildcard {
 				wildcardAllowed = true
 			} else {
-				users = append(users, user)
+				entities = append(entities, entity)
 			}
 		}
 
-		for _, user := range role.GetDatabaseUsers(types.Deny) {
-			if user == types.Wildcard {
+		for _, entity := range getter(role, types.Deny) {
+			if entity == types.Wildcard {
 				wildcardDenied = true
 			} else {
-				users = append(users, user)
+				entities = append(entities, entity)
 			}
 		}
 
 		result.wildcardDenied = result.wildcardDenied || wildcardDenied
 
-		if err := NewRoleSet(role).checkAccess(database, AccessMFAParams{Verified: true}); err == nil {
+		if err := NewRoleSet(role).checkAccess(r, AccessMFAParams{Verified: true}); err == nil {
 			result.wildcardAllowed = result.wildcardAllowed || wildcardAllowed
 		}
-
 	}
 
-	users = apiutils.Deduplicate(append(users, extraUsers...))
+	entities = apiutils.Deduplicate(append(entities, extraEntities...))
 
-	// check each individual user against the database.
-	for _, user := range users {
-		err := set.checkAccess(database, AccessMFAParams{Verified: true}, &DatabaseUserMatcher{User: user})
-		result.allowedDeniedMap[user] = err == nil
+	// check each individual entity against the database.
+	for _, entity := range entities {
+		err := set.checkAccess(r, AccessMFAParams{Verified: true}, matchMaker(entity))
+		result.allowedDeniedMap[entity] = err == nil
 	}
 
 	return result
 }
 
+// Check if a roleset is allowed to access a database.
+// This check is not exhaustive - it is more permissive than a real authorization check.
+// If db_user/db_name are blank, then we do not know what db_user/db_name will be used by default - A full RBAC check would deny blank db_user/db_name unless there is a wildcard allow rule.
+func (set RoleSet) CheckAccessToDatabase(dbRoute tlsca.RouteToDatabase, database types.Database) error {
```

Comment:
I'd recommend renaming this so it doesn't accidentally get used in RBAC in the future (since it is more permissive than RBAC checks).

Maybe something like:

```
// ShouldIssueDatabaseCert checks if a certificate should be issued
// for the specified database. Standard RBAC checks are still enforced
// at connection time - this check just attempts to "fail fast" with 
// a user-friendly error, preventing Teleport from issuing certs that
// will be denied on a subsequent connection attempt.
func (set RoleSet) ShouldIssueDatabaseCert(...
```

Reply:
I'm still not sure why this should be a part of RoleSet TBH. This is not an authorization check and like @zmb3 mentioned including this logic to the authz layer may only lead to confusion.

I look at this like a simple "CLI flag value validation" issue. tsh already knows all "allowed users", so why can't it just compare value in `--db-user` with the list of "allowed users" and error out if it's not in the list? This is a purely convenience check, I would not overthink this.

Reply:
We also already have a CheckAccessToResource function (or whatever it's called) so having another CheckAccessToDatabase will only increase confusion on why/when one is used versus the other.

Reply:
I moved it out of role set and into tsh as a `checkRoute` function. I elaborated on why the existing `checkAccess` function is not appropriate here. Now tsh will do thes "CLI validation" kind of checking before `tsh db login` or `tsh db connect`.